third_party/nvfuser/csrc/type.cpp

```diff
@@ -221,41 +136,48 @@ ValType promote_type(const ValType& t1, const ValType& t2) {
   TORCH_CHECK(false, "Expected promotable ValTypes but got: ", t1, " and ", t2);
 }
 
-static const char* data_type2string(DataType t) {
-  switch (t) {
-    case DataType::Bool:
-      return "bool";
-    case DataType::Double:
-      return "double";
-    case DataType::Float:
-      return "float";
-    case DataType::Half:
-      return "__half";
-    case DataType::BFloat16:
-      return "__bfloat";
-    case DataType::Int:
-      return "int64_t";
-    case DataType::Index:
-      return "nvfuser_index_t";
-    case DataType::Int32:
-      return "int";
-    case DataType::SMemAddress:
-      return "unsigned";
-    case DataType::ComplexFloat:
-      return "std::complex<float>";
-    case DataType::ComplexDouble:
-      return "std::complex<double>";
-    case DataType::Double_2:
-      return "Array<double, 2, 1>";
-    case DataType::Float_2:
-      return "Array<float, 2, 1>";
-    case DataType::Null:
-      return "null_type";
-    default:
-      break;
-  }
-  TORCH_INTERNAL_ASSERT(false, "No string found for data type.");
-  return nullptr;
+static std::string data_type2string(DataType t) {
+  return std::visit(
```

Comment:
If we just use the lambda, then `auto&& dtype` will be `std::variant<PrimDataType, ArrayOf. PointerOf>`, instead of any of the specific types. So we will not be able to do `switch(dtype)`, or `dtype.type`, etc. We need a dispatcher that does something like
```C++
if (std::holds_alternative<PrimDataType>(dtype) {
  PrimDataType t = std::get<PrimDataType>(dtype);
  ...
} else if (std::holds.....
```
And `std::visit` is designed to do this kind of dispatch. (actually my code is copy pasted from the example in https://en.cppreference.com/w/cpp/utility/variant/visit)

Reply:
TIL. Thanks.